src/tangents.jl

```diff
@@ -436,16 +441,88 @@ end
     tangent_field_zeros_exprs = ntuple(fieldcount(P)) do n
         if tangent_field_type(P, n) <: PossiblyUninitTangent
             V = PossiblyUninitTangent{tangent_type(fieldtype(P, n))}
-            return :(isdefined(x, $n) ? $V(zero_tangent(getfield(x, $n))) : $V())
+            return :(isdefined(x, $n) ? $V(zero_tangent_internal(getfield(x, $n))) : $V())
         else
-            return :(zero_tangent(getfield(x, $n)))
+            return :(zero_tangent_internal(getfield(x, $n)))
         end
     end
     backing_data_expr = Expr(:call, :tuple, tangent_field_zeros_exprs...)
     backing_expr = :($(backing_type(P))($backing_data_expr))
     return :($(tangent_type(P))($backing_expr))
 end
 
+@inline zero_tangent_internal(x::Union{Int8,Int16,Int32,Int64,Int128,IEEEFloat}, ::IdDict) = zero_tangent_internal(x)
+@inline function zero_tangent_internal(x::SimpleVector, stackdict::IdDict)
+    return map!(n -> zero_tangent_internal(x[n], stackdict), Vector{Any}(undef, length(x)), eachindex(x))
+end
+@inline function zero_tangent_internal(x::Array{P, N}, stackdict::IdDict) where {P, N}
+    if haskey(stackdict, x)
+        return stackdict[x]::tangent_type(typeof(x))
+    end
+    if isbitstype(P)
+        zt = _map_if_assigned!(Base.Fix2(zero_tangent_internal, stackdict), Array{tangent_type(P), N}(undef, size(x)...), x)
+        stackdict[x] = zt
+        return zt
+    else
+        zt = Array{tangent_type(P), N}(undef, size(x)...)
+        stackdict[x] = zt
+        for i in eachindex(x)
+            if isassigned(x, i)
+                zt[i] = zero_tangent_internal(x[i], stackdict)
+            end
+        end
+        return zt
+    end
+end
+function zero_tangent_internal(x::P, stackdict::IdDict) where {P<:Union{Tuple, NamedTuple}}
+    if tangent_type(P) == NoTangent
+        return NoTangent()
+    end
+    if haskey(stackdict, x)
```

Comment:
I think we can probably get away without doing this here. Since `Tuple`s and `NamedTuple`s are immutable, they don't have an address, so the same reasoning applies to them as does to `struct`s.

So I think you can avoid the check, and you don't ever need to put them in the `stackdict`, so the line just before the end of the function can be dropped as well.